### ISSUE TITLE
pytest-sugar is not working with pytest v5.4.0 (released on March 12th 2020)

### ISSUE THREAD

**Issue by MarcBoissonneault** [Teemu/pytest-sugar#187]
pytest-sugar is not working with pytest v5.4.0 (released on March 12th 2020)

```
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/_pytest/main.py", line 187, in wrap_session
INTERNALERROR>     config._do_configure()
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/_pytest/config/__init__.py", line 820, in _do_configure
INTERNALERROR>     self.hook.pytest_configure.call_historic(kwargs=dict(config=self))
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/pluggy/hooks.py", line 308, in call_historic
INTERNALERROR>     res = self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/pluggy/manager.py", line 93, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/pluggy/manager.py", line 87, in <lambda>
INTERNALERROR>     firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/pluggy/callers.py", line 208, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/pluggy/callers.py", line 80, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/pluggy/callers.py", line 187, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/pytest_sugar.py", line 176, in pytest_configure
INTERNALERROR>     sugar_reporter = SugarTerminalReporter(standard_reporter)
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/pytest_sugar.py", line 214, in __init__
INTERNALERROR>     self.writer = self._tw
INTERNALERROR> AttributeError: can't set attribute
```

**Comment by GuillaumeFavelier** [comment]
It seems that direct access to `writer` is deprecated and that `write()` should be used instead:

Reference:
https://docs.pytest.org/en/latest/deprecations.html#terminalreporter-writer

**Comment by blueyed** [comment]
@GuillaumeFavelier 
Your patch looks good, please consider creating a PR with it.

**Comment by donbeave** [comment]
Hello!

It looks like all fixes were merged already. Can I ask when the new version with fixes planned to be released?

**Comment by jakul** [comment]
Bump!

Can someone cut a release and put onto PyPi?

**Comment by WolfgangFellger** [comment]
https://github.com/Teemu/pytest-sugar/pull/188 hasn't even been merged, been sitting there approved for ten days now :-/

**Comment by RonnyPfannschmidt** [comment]
from personal experience i can say that a number of open-source maintainers currently have different priorities due to accommodating for stuff like daycare shut-down and corona preventions,

while i understand the unhappiness with delays,
lets have patience and lets hope and wish for the good health of the maintainers and their families

**Comment by blueyed** [comment]
Keep in mind though that this is basically a regression in pytest: it wanted to deprecate a property (out of nowhere TBH), and hasn't considered that some plugins like pytest-sugar might actually set this property.
Therefore (https://github.com/pytest-dev/pytest/issues/6951) it is more a pytest bug after all.

While this is unfortunate in general and I can see that parents have less time currently, it is still like ~2 weeks already without ttis being fixed.

IMHO:

1. there is not much reason for deprecating something just because
2. if it causes a regression it should be fixed quickly

pytest-sugar is not in the camp of having to fix the above after all: it should only have to adjust to DeprectionWarnings (which might still be fatail, but that is another issue) - an attempt to issue a DeprecationWarning that breaks things is a bug in the one that tried to issue it.  /cc @nicoddemus

**Comment by blueyed** [comment]
As mentioned in https://github.com/Teemu/pytest-sugar/pull/188#issuecomment-604462148 (8h ago) there is some patch available already, which is even in line what Ronny suggested.
It would be great if this could be fixed in pytest already.

**Comment by ssbarnea** [comment]
Sadly 3 weeks of being broken forced me to stop using pytest-sugar completely. 

I think that the responsibility in this area is a bit of a grey area but I can only recommend this: create scheduled jobs that run daily or at least weekly that run with pytest from master. This is what I do on several projects that highly dependent on a specific tool (being plugins or plugins like).

On the other hand it would be very useful if changes proposed on pytest would try to do a minimal compatibility testing with a set of maintained plugins, preventing changes that would fully broke them. Still this could be better done if these plugins would be hosted under pytest-dev organization, where I suppose it would be easier for a pytest core developer to patch a plugin that would be broken by an incoming (not merged yet) change made to pytest.

Note: full testing should probably not be necessary, but at least installation and running pytest collection or a very small set of test should be enough to evaluate that plugins do not break.

**Comment by truthdoug** [comment]
Reposting a comment I posted on the PR for this:

This will work to get pytest-sugar working until this has merged:

    pip install -e git+https://github.com/GuillaumeFavelier/pytest-sugar.git@b56fed42d5c3022ff507ab6ce81cda65a3a5f92a#egg=pytest-sugar

It simply installs the version from @GuillaumeFavelier's fork.

**Comment by RonnyPfannschmidt** [comment]
https://github.com/pytest-dev/pytest/pull/6986 will hopefully solve the pytest side, once i had a bit of time and peace of mind it was done very quickly

**Comment by igorbezr** [comment]
@truthdoug 
Thank you very mutch for this temporary solution !

@RonnyPfannschmidt 
Please clarify, it is correct that in pytest 5.4.2 this incompatibility with pytest-sugar hopefully will be fix ?

**Comment by RonnyPfannschmidt** [comment]
@igorbezr yes

**Comment by Lawouach** [comment]
Hello there,

I'm gettin ga little confused. What's the state of this? Are we waiting on pytest or this plugin to have a fix/release?

Cheers :)

**Comment by ankostis** [comment]
On the same page, is it possible for a fix this issue so that `pytest-sugar` is backward & future compatible to `pytest`?

**Comment by nicoddemus** [comment]
https://github.com/Teemu/pytest-sugar/pull/188 will make pytest-sugar use the recommended API, which is also backward compatible.

> Please clarify, it is correct that in pytest 5.4.2 this incompatibility with pytest-sugar hopefully will be fix ?

This is also true, `5.4.2` will also fix this.

**Comment by Lawouach** [comment]
Thanks! :)

Right so, until either of them are released, we should limit to pytests 5.4.1 or below, right?

**Comment by nicoddemus** [comment]
I think <5.4, as this was introduced in 5.4.0. 👍

**Comment by Lawouach** [comment]
great. thank you.

**Comment by MapleCCC** [comment]
> This will work to get pytest-sugar working until this has merged:
> 
> ```
> pip install -e git+https://github.com/GuillaumeFavelier/pytest-sugar.git@b56fed42d5c3022ff507ab6ce81cda65a3a5f92a#egg=pytest-sugar
> ```
> 
> It simply installs the version from @GuillaumeFavelier's fork.

For those who don't want to install in editable mode, you can drop the `-e` option when issuing the command propsed by @truthdoug .

**Comment by ankostis** [comment]
Just noting that as of today, latest pytest(5.4.1) & pytest-sugar(0.9.3) in PyPi still fail with:
```tb
$ pytest
Test session starts (platform: linux, Python 3.8.2, pytest 5.4.1, pytest-sugar 0.9.3)
rootdir: /project-dir, inifile: setup.cfg
plugins: cov-2.8.1, sphinx-0.2.2, sugar-0.9.3
collecting ... INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File ".../.venv/lib/python3.8/site-packages/_pytest/main.py", line 191, in wrap_session
INTERNALERROR>     session.exitstatus = doit(config, session) or 0
INTERNALERROR>   File ".../.venv/lib/python3.8/site-packages/_pytest/main.py", line 246, in _main
INTERNALERROR>     config.hook.pytest_collection(session=session)
INTERNALERROR>   File ".../.venv/lib/python3.8/site-packages/pluggy/hooks.py", line 286, in __call__
INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>   File ".../.venv/lib/python3.8/site-packages/pluggy/manager.py", line 93, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File ".../.venv/lib/python3.8/site-packages/pluggy/manager.py", line 84, in <lambda>
INTERNALERROR>     self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
INTERNALERROR>   File ".../.venv/lib/python3.8/site-packages/pluggy/callers.py", line 208, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File ".../.venv/lib/python3.8/site-packages/pluggy/callers.py", line 80, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File ".../.venv/lib/python3.8/site-packages/pluggy/callers.py", line 187, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File ".../.venv/lib/python3.8/site-packages/_pytest/main.py", line 257, in pytest_collection
INTERNALERROR>     return session.perform_collect()
INTERNALERROR>   File ".../.venv/lib/python3.8/site-packages/_pytest/main.py", line 454, in perform_collect
INTERNALERROR>     hook.pytest_collection_modifyitems(
INTERNALERROR>   File ".../.venv/lib/python3.8/site-packages/pluggy/hooks.py", line 286, in __call__
INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>   File ".../.venv/lib/python3.8/site-packages/pluggy/manager.py", line 93, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File ".../.venv/lib/python3.8/site-packages/pluggy/manager.py", line 84, in <lambda>
INTERNALERROR>     self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
INTERNALERROR>   File ".../.venv/lib/python3.8/site-packages/pluggy/callers.py", line 208, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File ".../.venv/lib/python3.8/site-packages/pluggy/callers.py", line 80, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File ".../.venv/lib/python3.8/site-packages/pluggy/callers.py", line 187, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File ".../.venv/lib/python3.8/site-packages/_pytest/mark/__init__.py", line 151, in pytest_collection_modifyitems
INTERNALERROR>     deselect_by_mark(items, config)
INTERNALERROR>   File ".../.venv/lib/python3.8/site-packages/_pytest/mark/__init__.py", line 145, in deselect_by_mark
INTERNALERROR>     config.hook.pytest_deselected(items=deselected)
INTERNALERROR>   File ".../.venv/lib/python3.8/site-packages/pluggy/hooks.py", line 286, in __call__
INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>   File ".../.venv/lib/python3.8/site-packages/pluggy/manager.py", line 93, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File ".../.venv/lib/python3.8/site-packages/pluggy/manager.py", line 84, in <lambda>
INTERNALERROR>     self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
INTERNALERROR>   File ".../.venv/lib/python3.8/site-packages/pluggy/callers.py", line 208, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File ".../.venv/lib/python3.8/site-packages/pluggy/callers.py", line 80, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File ".../.venv/lib/python3.8/site-packages/pluggy/callers.py", line 187, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File ".../.venv/lib/python3.8/site-packages/_pytest/terminal.py", line 431, in pytest_deselected
INTERNALERROR>     self._add_stats("deselected", items)
INTERNALERROR>   File ".../.venv/lib/python3.8/site-packages/_pytest/terminal.py", line 402, in _add_stats
INTERNALERROR>     self._set_main_color()
INTERNALERROR>   File ".../.venv/lib/python3.8/site-packages/_pytest/terminal.py", line 1090, in _set_main_color
INTERNALERROR>     self._main_color = self._determine_main_color(bool(unknown_types))
INTERNALERROR>   File ".../.venv/lib/python3.8/site-packages/_pytest/terminal.py", line 1077, in _determine_main_color
INTERNALERROR>     elif "passed" in stats or not self._is_last_item:
INTERNALERROR>   File ".../.venv/lib/python3.8/site-packages/_pytest/terminal.py", line 496, in _is_last_item
INTERNALERROR>     return len(self._progress_nodeids_reported) == self._session.testscollected
INTERNALERROR> AttributeError: 'NoneType' object has no attribute 'testscollected'


Results (0.57s):
      63 deselected
```

**Comment by jakul** [comment]
I've raised https://github.com/Teemu/pytest-sugar/issues/194 to track that error - this one is fixed
